src/compartment.c

```diff
@@ -639,65 +427,452 @@ comp_exec(
+            lib_dep->lib_dep_names = realloc(lib_dep->lib_dep_names,
+                (lib_dep->lib_dep_count + 1) * sizeof(char *));
+            lib_dep->lib_dep_names[lib_dep->lib_dep_count]
+                = malloc(strlen(&dynstr_tbl[dyn_entries[i].d_un.d_val]));
+            strcpy(lib_dep->lib_dep_names[lib_dep->lib_dep_count],
+                &dynstr_tbl[dyn_entries[i].d_un.d_val]);
+            lib_dep->lib_dep_count += 1;
         }
-        free(ld->lib_segs);
-        for (size_t j = 0; j < ld->lib_syms_count; ++j)
+    }
+
+    free(dynstr_tbl);
+    free(dyn_entries);
+}
+
+static void
+find_comp_entry_points(
+    char **entry_points, size_t entry_point_count, struct Compartment *new_comp)
+{
+    new_comp->entry_points
+        = malloc(entry_point_count * sizeof(struct CompEntryPoint));
+    for (size_t i = 0; i < entry_point_count; ++i)
+    {
+        struct LibSymSearchResult found_sym
+            = find_lib_dep_sym_in_comp(entry_points[i], new_comp);
+        if (found_sym.lib_idx == USHRT_MAX)
         {
-            free(ld->lib_syms[j].sym_name);
+            errx(1, "Did not find entry point %s!\n", entry_points[i]);
         }
-        free(ld->lib_syms);
-        free(ld);
+        struct CompEntryPoint new_entry_point
+            = { entry_points[i], extract_sym_offset(new_comp, found_sym) };
+        new_comp->entry_points[new_comp->entry_point_count] = new_entry_point;
+        new_comp->entry_point_count += 1;
     }
-    free(to_clean->lib_deps);
-    free(to_clean->intercept_patches);
+}
 
-    free(to_clean);
-    // TODO
+static void
+find_comp_intercepts(char **intercepts, void **intercept_addrs,
+    size_t intercept_count, struct Compartment *new_comp)
+{
+    // Find symbols for intercepts
+    char **intercept_names = malloc(intercept_count * sizeof(char *));
+    const char *so_plt_suffix = "@plt";
+    for (size_t i = 0; i < intercept_count; ++i)
+    {
+        size_t to_intercept_name_len
+            = strlen(intercepts[i]) + strlen(so_plt_suffix) + 1;
+        intercept_names[i] = malloc(to_intercept_name_len);
+        strcpy(intercept_names[i], intercepts[i]);
+        strcat(intercept_names[i], so_plt_suffix);
+    }
+    for (size_t i = 0; i < intercept_count; ++i)
+    {
+        struct LibSymSearchResult found_sym
+            = find_lib_dep_sym_in_comp(intercept_names[i], new_comp);
+        if (found_sym.lib_idx == USHRT_MAX)
+        {
+            continue;
+        }
+
+        // TODO double check
+        comp_add_intercept(new_comp,
+            (uintptr_t) extract_sym_offset(new_comp, found_sym),
+            (uintptr_t) intercept_addrs[i]);
+        free(intercept_names[i]);
+    }
+    free(intercept_names);
+}
+
+static void
+resolve_rela_syms(struct Compartment *new_comp)
+{
+    // Find all symbols for eager relocation mapping
+    for (size_t i = 0; i < new_comp->libs_count; ++i)
+    {
+        for (size_t j = 0; j < new_comp->libs[i]->rela_maps_count; ++j)
+        {
+
```

Comment:
Extraneous blank line.

Reply:
Removed in 41fb04c639efcf1adac5d450557af8d1226654ef.